### PR TITLE
Speed up findingDetectors for IPeaks in PredictFractionalPeaks

### LINKED ISSUE #22060: PredictFractionalPeaks speedup
Pascal has noted that `PredictFractionalPeaks` algorithm is running slowly.
This algorithm investigates all hkl space so the main slow down is there (3 nested for loops).
It would be good to see if we can increase the speed of this.

### PR DESCRIPTION
**Pull request by ElliotAOram**
PredictFractionalPeaks was using a new raytracer for every peak it created and therefore throwing away the caching information about the bounding boxs of detectors / panels causing a huge slow down. Resolved by exposing the findDetector(raytracer) overload to IPeaks and using single raytracer for all calls.
Work also include small refactor for const autos

Description of work.

**To test:**

* Pascals script at ISIS should be a good test for this:
```
LoadRaw(Filename='WISH00040151.raw', OutputWorkspace='40151', LoadLogFiles=False)
CropWorkspace(InputWorkspace='40151', OutputWorkspace='40151', XMin=6000, XMax=99000)
ConvertUnits(InputWorkspace='40151', OutputWorkspace='40151', Target='dSpacing')
FindSXPeaks(InputWorkspace='40151', PeakFindingStrategy='AllPeaks', AbsoluteBackground=10, ResolutionStrategy='AbsoluteResolution', XResolution=0.10000000000000001, PhiResolution=2, TwoThetaResolution=2, OutputWorkspace='40151_peaks')
FindUBUsingLatticeParameters(PeaksWorkspace='40151_peaks', a=10.01, b=8.9100000000000001, c=5.5099999999999998, alpha=90, beta=107.51000000000001, gamma=90, Tolerance=0.20000000000000001, Iterations=5)
PredictPeaks(InputWorkspace='40151_peaks', WavelengthMin=0.5, WavelengthMax=10, ReflectionCondition='C-face centred', OutputWorkspace='40151_predicted')
PredictFractionalPeaks(Peaks='40151_predicted', FracPeaks='40151_predicted_frac', HOffset='0.6,-0.6', LOffset='-0.08,0.08', IncludeAllPeaksInRange=True, Hmin=-2, Hmax=2, Kmin=-2, Kmax=2, Lmin=-2, Lmax=2)
```

* Run the script and ensure that `PredictFractionalPeaks` is fast (~10secs) previously on release build this was running at around 2mins so ensure there is a speed up in this alg. 

Fixes #22060

**Release Notes** 
This bug was not in the last release (introduced as a by-product of instrument changes this release) hence no need to add it to the release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by OwenArnold — CHANGES_REQUESTED** [review]
I'm currently building/testing on my platform now, but the fix looks sensible.

If this makes things faster than release 3.11, I would suggest adding a not about it to the release notes.

Also, and this could be done as a separate PR, could you please add at least one performance test for this algorithm so that we can track future regressions. I very much think that Pascal would not like to see this ever slow down again.

File: Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp [hunk 1]
```diff
@@ -89,7 +89,8 @@ void export_IPeak() {
            ":class:`~mantid.geometry.Goniometer` rotation was NOT taken "
            "out.\n"
            "Note: There is no 2*pi factor used, so \\|Q| = 1/wavelength.")
-      .def("findDetector", &IPeak::findDetector, arg("self"),
+      .def("findDetector", (bool (IPeak::*)()) & IPeak::findDetector,
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp]: @OwenArnold I think this was already exposed and the changes here just ensure the previous overload is picked?

**Review by OwenArnold — APPROVED** [review]

**Comment by OwenArnold** [comment]
Prior to changes took 1 min 25 seconds after changes takes under 1 second. Haven't compared to release 3.11, but as mentioned above, if better speed is achieved, it should be mentioned in release notes.

**Comment by martyngigg** [comment]
@OwenArnold I ran the last release and it's pretty similar so probably not worth mentioning.

**Comment by OwenArnold** [comment]
@martyngigg. Yes, no point mentioning it. We should still add a performance test for this algorithm, but as said before, can be done as a separate issue.

**Pull request merged**